ako-gateway-api/k8s/ako_init.go

```diff
@@ -451,3 +456,169 @@ func (c *GatewayController) cleanupStaleVSes() {
+					c.DeleteModels()
+					SetDeleteSyncChannel()
+
+				} else {
+					status.NewStatusPublisher().ResetStatefulSetAnnotation(status.GatewayObjectDeletionStatus)
+					akogatewayapilib.AKOControlConfig().PodEventf(corev1.EventTypeNormal, lib.AKODeleteConfigUnset, "DeleteConfig unset in configmap, sync would be enabled")
+					quickSyncCh <- struct{}{}
+				}
+			}
+
+		},
+		DeleteFunc: func(obj interface{}) {
+			utils.AviLog.Warnf("avi k8s configmap deleted, shutting down api server")
+		},
+	}
+
+	c.informers.ConfigMapInformer.Informer().AddEventHandler(configMapEventHandler)
+	go c.informers.ConfigMapInformer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh,
+		c.informers.ConfigMapInformer.Informer().HasSynced,
+	) {
+		runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+	} else {
+		utils.AviLog.Infof("Caches synced")
+	}
+	return nil
+}
+
+func DeleteConfigFromConfigmap(cs kubernetes.Interface) (bool, error) {
+	cmNS := utils.GetAKONamespace()
+	cm, err := cs.CoreV1().ConfigMaps(cmNS).Get(context.TODO(), lib.AviConfigMap, metav1.GetOptions{})
+	if err == nil {
+		return delConfigFromData(cm.Data), err
+	}
+	utils.AviLog.Warnf("error while reading configmap, sync would be disabled: %v", err)
+	return true, err
+}
+func delConfigFromData(data map[string]string) bool {
+	var delConf bool
+	if val, ok := data[lib.DeleteConfig]; ok {
+		if val == "true" {
+			utils.AviLog.Infof("deleteConfig set in configmap, sync would be disabled")
+			akogatewayapilib.AKOControlConfig().PodEventf(corev1.EventTypeNormal, lib.AKODeleteConfigSet, "DeleteConfig set in configmap, sync would be disabled")
+			delConf = true
+		}
+	}
+	lib.SetDeleteConfigMap(delConf)
+	return delConf
+}
+
+// DeleteModels : Delete models and add the model name in the queue.
+// The rest layer would pick up the model key and delete the objects in Avi
+func (c *GatewayController) DeleteModels() {
+	utils.AviLog.Infof("Deletion of all avi objects triggered")
+	publisher := status.NewStatusPublisher()
+	publisher.AddStatefulSetAnnotation(status.GatewayObjectDeletionStatus, lib.ObjectDeletionStartStatus)
+	allModels := objects.SharedAviGraphLister().GetAll()
+	allModelsMap := allModels.(map[string]interface{})
+	if len(allModelsMap) == 0 {
+		utils.AviLog.Infof("No Avi Object to delete, status would be updated in Statefulset")
+		publisher.AddStatefulSetAnnotation(status.GatewayObjectDeletionStatus, lib.ObjectDeletionDoneStatus)
+		return
+	}
+	sharedQueue := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
+	for modelName := range allModelsMap {
+		objects.SharedAviGraphLister().Save(modelName, nil)
+		bkt := utils.Bkt(modelName, sharedQueue.NumWorkers)
+		utils.AviLog.Infof("Deleting objects for model: %s", modelName)
+		//graph queue prometheus
+		sharedQueue.Workqueue[bkt].AddRateLimited(modelName)
+	}
+}
+
+func SetDeleteSyncChannel() {
+	// Wait for maximum 30 minutes for the sync to get completed
+	if lib.ConfigDeleteSyncChan == nil {
```

Comment:
@aaha97: Should we use different `ConfigDeleteSyncChan` for GW and AKO container. Reason: IF one is done and other is pending, event will generate on this channel which will be catched by `select` loop and it will generate an event and annotation update even though it is not completed cleanup task .